src/folding/circuits/hypernova/utils.rs

```diff
@@ -0,0 +1,129 @@
+use crate::folding::circuits::utils::VecFpVar;
+use ark_ff::PrimeField;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    fields::{fp::FpVar, FieldVar},
+    ToBitsGadget,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+use std::{borrow::Borrow, marker::PhantomData};
+
+/// `GammaVar` is a wrapper around an `FpVar` for $\gamma$.
+#[derive(Clone)]
+pub struct GammaVar<F: PrimeField> {
+    pub value: FpVar<F>,
+}
+
+impl<F: PrimeField> AllocVar<F, F> for GammaVar<F> {
+    fn new_variable<T: Borrow<F>>(
+        cs: impl Into<Namespace<F>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        FpVar::new_variable(cs, f, mode).map(|value| Self { value })
+    }
+}
+
+impl<F: PrimeField> GammaVar<F> {
+    /// An helper to compute $\gamma^{pow}$
+    pub fn pow(&self, pow: FpVar<F>) -> Result<FpVar<F>, SynthesisError> {
+        let pow = pow.to_bits_le()?;
+        self.value.pow_le(&pow)
+    }
+}
+
+/// Gadget to compute the sum of all $\gamma^{j} \cdot eq(r_{x_j}, r_x^{\prime}) \cdot \sigma_j$.
+pub struct SumMulsGammaPowsEqSigmaGadget<F: PrimeField> {
+    _f: PhantomData<F>,
+}
+
+impl<F: PrimeField> SumMulsGammaPowsEqSigmaGadget<F> {
+    /// Computes the sum $\Sigma_{j}^{j + n} \gamma^{j} \cdot eq_eval \cdot \sigma_{j}$, where $n$ is the length of the `sigmas` vector
+    /// It corresponds to the first term of the sum that $\mathcal{V}$ has to compute at section 5, step 5 of "A multi-folding scheme for CCS".
+    ///
+    /// # Arguments
+    /// - `sigmas`: vector of $\sigma_j$ values
+    /// - `eq_eval`: the value of $\tilde{eq}(x_j, x^{\prime})$
+    /// - `gamma`: a `GammaVar`, which supports a `pow` method, representing $\gamma$
+    /// - `j`: the power at which we start to compute $\gamma^{j}$. This is needed in the contexxt of multifolding.
+    ///
+    /// # Notes
+    /// In the context of multifolding, `j` corresponds to `ccs.t` in `compute_c_from_sigmas_and_thetas`
+    pub fn sum_muls_gamma_pows_eq_sigma(
+        sigmas: VecFpVar<F>,
+        eq_eval: FpVar<F>,
+        gamma: GammaVar<F>,
+        j: FpVar<F>,
+    ) -> Result<FpVar<F>, SynthesisError> {
+        let mut result = FpVar::<F>::zero();
+        let mut gamma_pow = gamma.pow(j).map_err(|_| SynthesisError::Unsatisfiable)?;
+        for sigma in sigmas {
+            result += gamma_pow.clone() * eq_eval.clone() * sigma;
+            gamma_pow *= gamma.value.clone();
+        }
+        Ok(result)
+    }
```

Comment:
I think that the `sum_muls_gamma_pows_eq_sigma` method can be done without the `GammaVar` type wraper, being easier to follow, and getting rid of the GammaVar related code (lines 11 to 33). How would you see it?

It would be something like this, plus removing the lines from 11 to 33:
```suggestion
    pub fn sum_muls_gamma_pows_eq_sigma(
        sigmas: VecFpVar<F>,
        eq_eval: FpVar<F>,
        gamma: FpVar<F>,
        j: FpVar<F>,
    ) -> Result<FpVar<F>, SynthesisError> {
        let mut result = FpVar::<F>::zero();
        let mut gamma_pow = gamma.pow_le(&j.to_bits_le()?)?;
        for sigma in sigmas {
            result += gamma_pow.clone() * eq_eval.clone() * sigma;
            gamma_pow *= gamma.clone();
        }
        Ok(result)
    }
```


(For the pow logic, is like when we do scalar multiplication inside a circuit, we do `point.scalar_mul_le(bits.iter()?)`, we could add a type wrapper for that and add custom methods, but since it can already be done with a single line I think that it is easier to follow rather than adding extra calls to custom methods to extra type wrappers.)

Reply:
[Yup](https://github.com/privacy-scaling-explorations/folding-schemes/pull/53/commits/e3fcb8790d790c25267fcdfb40e820bf76a6aa3e)!